### PR TITLE
Handle new APIML unique cookie identifier

### PR DESCRIPTION
**Pull request by zFernand0**
**What It Does**

- resolves https://github.com/zowe/imperative/issues/979
- ⚠️ depends on https://github.com/zowe/imperative/pull/996 ⚠️ 
- allowed dynamic APIML token types on various commands
- prevented basepath requirement error on invalid credentials or invalid/expired token
- allowed `logout` operation with invalid and/or expired tokens

**How to Test**

**Note**: you may need to build the APIML (https://github.com/zowe/api-layer) locally in order to configure it to use unique cookie identifiers

Commands to try:
- `zowe auth login apiml`
- `zowe auth logout apiml`
- `zowe config secure`
- `zowe config auto-init`

---

Scenarios from Imperative:
- prevented multiple logout operations from failing due to not having a token available
  - Scenarios: 
    - have an expired token in your vault, or 
    - save a dummy value for the tokenValue property, or 
    - have an empty string as your token
  - Before:   
     ```
      λ zowe auth logout apiml
      Enter the host name of your service: sys1.com
      Enter the port number for your service: 10010
      Command Error:
      This operation requires authentication.

      z/OSMF REST API Error:
      Rest API failure with HTTP(S) status 401
      messages:
        -
          messageType:    ERROR
          messageNumber:  ZWEAT100E
          messageContent: Token is expired for URL '/api/v1/gateway/auth/logout'
          messageKey:     org.zowe.apiml.security.expiredToken

      Host:      sys1.com
      Port:      10010
      Base Path:
      Resource:  /api/v1/gateway/auth/logout
      Request:   POST
      Headers:   undefined
      Payload:   POST

      Error Details:
      Token is not valid or expired.

      For CLI usage, see `zowe auth login apiml --help`
      ```
  - After:
    ```
    # zdev auth logout apiml
    Logout successful. The authentication token has been revoked and removed from your 'base_real' base profile
    ```
- added support for multiple token authentication processes in a single `config secure` command
  - Scenarios:
    - with a config file that hase more than one base profile with `tokenValue` in the secure array
  - Before:
    ```
    # zdev config secure --gc
    Enter profiles.base_real.properties.tokenValue - blank to skip: 
    Enter profiles.base_mock.properties.tokenValue - blank to skip:
    ```
  - After:
    ```
    # zdev config secure --gc
    Processing secure properties for profile: base_mock
    Logging in to your API Mediation Layer
    Enter the user name for your API Mediation Layer (will be hidden): 
    Enter the password for your API Mediation Layer (will be hidden): 
    Processing secure properties for profile: base_real
    Logging in to your API Mediation Layer
    Enter the user name for your API Mediation Layer (will be hidden): 
    Enter the password for your API Mediation Layer (will be hidden): 
    ```
  - added an entry to the console.log to inform which profile credentials are being prompted for
    - Notice the `Processing secure properties for profile: <profile_name>` above
- prevented auto-init from performing two login operations on a single command
  - Scenarios:
    - perform a `zowe auth login apiml` operation and then remove the token from the vaullt
  - Before:
    - 1st login when prompted for the `tokenValue` (which will ask for `user:password` to login)
    - 2nd login in the `doAutoInit` functionality since the `user:password` are still provided in the ISession
  - After:
    - Only one login operation

---

Reverted scenario:
- ~allowed logout operation to remove token type and/or token value when either is not specified in the config file~
  - NOTE: this scenario conflicts with the ability to logout a token different from the one stored in the vault
  - Scenarios:
    - with a config that's missing either `tokenType` or `tokenValue` 
  - Before:
      ```
      λ zowe auth logout apiml
      Enter the host name of your service: sys1.com
      Enter the port number for your service: 10010
      Command Error:
      This operation requires authentication.

      z/OSMF REST API Error:
      Rest API failure with HTTP(S) status 401
      messages:
        -
          messageType:    ERROR
          messageNumber:  ZWEAT100E
          messageContent: Token is expired for URL '/api/v1/gateway/auth/logout'
          messageKey:     org.zowe.apiml.security.query.tokenNotProvided

      Host:      sys1.com
      Port:      10010
      Base Path:
      Resource:  /api/v1/gateway/auth/logout
      Request:   POST
      Headers:   undefined
      Payload:   POST

      Error Details:
      Token is not valid or expired.

      For CLI usage, see `zowe auth login apiml --help`
      ```
  - After:
    ```
    # zdev auth logout apiml
    Logout successful. The authentication token has been revoked.
    ```

---

**Review Checklist**
I certify that I have:
- [x] tested my changes
- [ ] added/updated automated tests
- [ ] updated the changelog
- [x] followed the [contribution guidelines](https://github.com/zowe/zowe-cli/blob/master/CONTRIBUTING.md)

**Additional Comments**
- related issues: 
  - https://github.com/zowe/zowe-cli/issues/1341
  - https://github.com/zowe/zowe-cli/issues/1650
  - https://github.com/zowe/zowe-cli/issues/1041

### CODE REVIEW

**Review by gejohnston — APPROVED** [review]
Looks good.

File: packages/cli/src/config/auto-init/ApimlAutoInitHandler.ts [hunk 1]
```diff
@@ -78,7 +78,7 @@ export default class ApimlAutoInitHandler extends BaseAutoInitHandler {
             session.ISession.tokenValue = await Login.apimlLogin(session);
             session.ISession.storeCookie = false;
             session.ISession.type = SessConstants.AUTH_TYPE_TOKEN;
-            session.ISession.user = session.ISession.password = session.ISession.cert = session.ISession.certKey = undefined;
+            session.ISession.base64EncodedAuth = session.ISession.user = session.ISession.password = session.ISession.cert = session.ISession.certKey = undefined;
```
Comment (awharn) [inline packages/cli/src/config/auto-init/ApimlAutoInitHandler.ts:81]: Linter says this line is too long.

**Review by awharn — COMMENTED** [review]
One small observation

File: packages/cli/src/config/auto-init/ApimlAutoInitHandler.ts [hunk 2]
```diff
@@ -73,12 +73,17 @@ export default class ApimlAutoInitHandler extends BaseAutoInitHandler {
         if ((session.ISession.user && session.ISession.password) || (session.ISession.cert && session.ISession.certKey)) {
             // If it is basic authentication, we need to set the auth type.
             if (session.ISession.password) { session.ISession.type = SessConstants.AUTH_TYPE_TOKEN; }
-            session.ISession.tokenType = SessConstants.TOKEN_TYPE_APIML;
+            if (session.ISession.tokenType == null) {
```
Comment (awharn) [inline packages/cli/src/config/auto-init/ApimlAutoInitHandler.ts:76]: What if the tokenType is not APIML?

Comment (zFernand0) [reply packages/cli/src/config/auto-init/ApimlAutoInitHandler.ts:76]: The `zowe config auto-init` command should only work with APIML tokens. I don't believe there should be an instance where the tokenType should be different. It could be dynamic, in which case it will get replaced after the login operation is performed further down.

Comment (awharn) [reply packages/cli/src/config/auto-init/ApimlAutoInitHandler.ts:76]: I was just wondering in case something else (unlikely) was setting up tokenType in the base profile as something else, and getting caught up with that value during the config load. Just an edge case, not sure if it makes sense in this context.

File: packages/core/__tests__/auth/__unit__/Logout.apiml.unit.test.ts [hunk 3]
```diff
@@ -69,28 +71,56 @@ describe("Auth Logout APIML unit tests", () => {
...
+                    caughtError = error;
+                }
+                expect(caughtError).toBeDefined();
+                expect(caughtError.message).toContain(realErrorText);
+                expect(caughtError instanceof ImperativeError).toEqual(true);
+                return caughtError;
+            };
+            // Token is invalid (logged out bu tnot expired)
+            expect(await runTest("org.zowe.apiml.security.query.invalidToken")).toBeDefined();
+            // Token is expored (old token)
```
Comment (awharn) [inline packages/core/__tests__/auth/__unit__/Logout.apiml.unit.test.ts:102]: expored? :P

File: packages/core/__tests__/auth/__unit__/Logout.apiml.unit.test.ts [hunk 4]
```diff
@@ -69,28 +71,56 @@ describe("Auth Logout APIML unit tests", () => {
...
+                    await Logout.apimlLogout(fakeSession);
+                } catch (error) {
+                    caughtError = error;
+                }
+                expect(caughtError).toBeDefined();
+                expect(caughtError.message).toContain(realErrorText);
+                expect(caughtError instanceof ImperativeError).toEqual(true);
+                return caughtError;
+            };
+            // Token is invalid (logged out bu tnot expired)
```
Comment (awharn) [inline packages/core/__tests__/auth/__unit__/Logout.apiml.unit.test.ts:100]: but not

File: packages/core/__tests__/auth/__unit__/__snapshots__/Login.apiml.unit.test.ts.snap [hunk 5]
```diff
@@ -8,7 +8,7 @@ REST API Failure with HTTP(S) status 401
 
 Host:      undefined
 Port:      undefined
-Base Path: undefined
+Base Path:
```
Comment (awharn) [inline packages/core/__tests__/auth/__unit__/__snapshots__/Login.apiml.unit.test.ts.snap:11]: This doesn't seem quite right. If login fails with a 401, are we expecting the base path to be a space or similar, instead of undefined?

**Review by awharn — COMMENTED** [review]
Just a couple more small questions/comments.

File: packages/cli/__tests__/auth/__integration__/__snapshots__/cli.auth.apiml.integration.test.ts.snap [hunk 6]
```diff
@@ -151,7 +151,7 @@ exports[`auth login/logout apiml help should display the logout help 1`] = `
       The type of token to get and use for the API. Omit this option to use the
       default token type, which is provided by 'zowe auth login'.
 
-      Allowed values: apimlAuthenticationToken, jwtToken, LtpaToken2
+      Allowed values: ^apimlAuthenticationToken.*, jwtToken, LtpaToken2
```
Comment (t1m0thyj) [inline packages/cli/__tests__/auth/__integration__/__snapshots__/cli.auth.apiml.integration.test.ts.snap:154]: Do we want to show regexes like this in the CLI help output? I think users are more likely to be familiar with wildcards which are simpler to read (e.g., `apimlAuthenticationToken*`)

I'm just curious about this, not requesting a change - because I assume it would require enhancing Imperative to support displaying "allowed values" in a different format.

File: packages/core/src/auth/Logout.ts [hunk 7]
```diff
@@ -25,24 +25,52 @@ export class Logout {
...
+                        default:
+                            errorToThrow = null;
+                            break;
+                    }
+                }
+            }
+
+            if (errorToThrow) {
+                throw errorToThrow;
+            }
```
Comment (t1m0thyj) [inline packages/core/src/auth/Logout.ts:70]: Where possible I think we should prefer checking HTTP error codes (well-defined behavior) over looking for specific substrings in the error message (subject to change).

I think we discussed IRL that it may be possible to replace the "V2_TOKEN" checks with a single check for error code 401 🙂 But maybe I am confused or remembering wrong?

Comment (zFernand0) [reply packages/core/src/auth/Logout.ts:70]: added a note for clarity (and the block of code that we could replace all of the `V2_TOKEN_...` checks with) 😋

File: packages/cli/__tests__/auth/__integration__/__snapshots__/cli.auth.apiml.integration.test.ts.snap (hunk 6 above)
Comment (zFernand0) [reply packages/cli/__tests__/auth/__integration__/__snapshots__/cli.auth.apiml.integration.test.ts.snap:154]: I'm afraid it would require a change in imperative to provide a nicer display for `allowableValues` 😋 
I don't mind opening an issue for this though 😋

File: packages/cli/src/config/auto-init/ApimlAutoInitHandler.ts [hunk 8]
```diff
@@ -99,27 +103,55 @@ export default class ApimlAutoInitHandler extends BaseAutoInitHandler {
...
+                type: this.mProfileType,
+                properties: {
+                    ...config.api.profiles.get(activeBaseProfile),
+                    host: session.ISession.hostname ?? params.arguments.host,
+                    port: session.ISession.port ?? params.arguments.port,
+                    rejectUnauthorized: session.ISession.rejectUnauthorized ?? params.arguments.rejectUnauthorized
+                },
+                secure: []
+            });
+        }
```
[the remark this answers is not in the document]
Comment (zFernand0) [reply packages/cli/src/config/auto-init/ApimlAutoInitHandler.ts:134]: I'm afraid this is just to keep the command behaving the same way as before.
in the scenario where there was no previous config, we were not storing these properties.
And when there was an existing base profile... the properties would have been properly set in the `ISession` object. 

I could remove the `?? params.arguments...` since I do not believe this will ever be needed 😋

File: packages/core/__tests__/auth/__unit__/__snapshots__/Login.apiml.unit.test.ts.snap (hunk 5 above)
Comment (t1m0thyj) [reply packages/core/__tests__/auth/__unit__/__snapshots__/Login.apiml.unit.test.ts.snap:11]: In the ZosmfRestClient error handling, could we add a check for `this.session.ISession.tokenValue != null` on error messages for invalid token? This might make the base path trickery unnecessary 😋

**Review by t1m0thyj — APPROVED** [review]
LGTM, thanks @zFernand0!

**Review by gejohnston — APPROVED** [review]
LGTM.

**Review by traeok — APPROVED** [review]
LGTM, thanks @zFernand0 😃
